### PR TITLE
#8133 - Impossible to restrict adding invalid monomer types (e.g. Unknown base, Micromolecule) through API

### LINKED ISSUE #8133: Impossible to restrict adding invalid monomer types (e.g. Unknown base, Micromolecule) through API
**Description:**
Ketcher allows adding monomers with modification types that should not be available in the Library.
Currently, entities such as Unknown base, Micromolecule, and other system or ambiguous modification types can be added through the API using
ketcher.updateMonomersLibrary(sdfString, { format: 'sdf' }).
These entities appear in the Library UI, although they are not expected to be visible or usable.

**Steps to Reproduce**
1. Open Ketcher in Macromolecule mode.
2. Execute one of the following API commands in the browser console:
**Example 1 — Unknown base:**
```
await ketcher.updateMonomersLibrary(`
  -INDIGO-1009299991D
  0  0  0  0  0  0  0  0  0  0  0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 1 0 0 0 0
M  V30 BEGIN ATOM
M  V30 1 XUB 11.875 -6.75 0.0 0 CLASS=BASE
M  V30 END ATOM
M  V30 END BOND
M  V30 END CTAB
M  V30 BEGIN TEMPLATE
M  V30 TEMPLATE 1 BASE/XUB/XUB NATREPLACE=BASE/A
M  V30 BEGIN CTAB
M  V30 COUNTS 12 12 1 0 0
M  V30 BEGIN ATOM
M  V30 1 N 0.000 0.000 0.0 0
M  V30 2 C 1.200 0.000 0.0 0
M  V30 3 N 2.000 0.800 0.0 0
M  V30 4 C -1.200 0.000 0.0 0
M  V30 5 N -1.800 0.800 0.0 0
M  V30 6 N -1.800 -0.800 0.0 0
M  V30 7 H 2.200 1.800 0.0 0
M  V30 8 C 1.400 1.800 0.0 0
M  V30 9 N 0.200 1.800 0.0 0
M  V30 10 H 2.400 -0.800 0.0 0
M  V30 11 H -2.400 -0.800 0.0 0
M  V30 12 H 0.000 -1.800 0.0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 2 2 3
M  V30 3 1 3 8
M  V30 4 1 8 9
M  V30 5 1 1 9
M  V30 END BOND
M  V30 BEGIN SGROUP
M  V30 1 SUP 1 ATOMS=(1 1 2 3 4 5 6 7 8 9 10 11 12) XBONDS=(1 12) LABEL=XUB CLASS=BASE
M  V30 END SGROUP
M  V30 END CTAB
M  V30 END TEMPLATE
M  END
>  <aliasHELM>
XUB

>  <idtAliases>
base=xubBase,ep5=xubEp5,ep3=xubEp3,i=xubI

>  <modificationType>
Unknown base

$$$$
`, { format: 'sdf' })
```

**Example 2 — Micromolecule:**
```
await ketcher.updateMonomersLibrary(`
  -INDIGO-10102512001D
  0  0  0  0  0  0  0  0  0  0  0 V3000
M  V30 BEGIN CTAB
M  V30 COUNTS 1 0 0 0 0
M  V30 BEGIN ATOM
M  V30 1 MCM 11.875 -6.75 0.0 0 CLASS=LINKER
M  V30 END ATOM
M  V30 END BOND
M  V30 END CTAB
M  V30 BEGIN TEMPLATE
M  V30 TEMPLATE 1 LINKER/MCM88/MCM88 NATREPLACE=LINKER/GEN
M  V30 BEGIN CTAB
M  V30 COUNTS 6 6 1 0 0
M  V30 BEGIN ATOM
M  V30 1 C 0.000 0.000 0.0 0
M  V30 2 C 1.200 0.000 0.0 0
M  V30 3 O 2.000 0.800 0.0 0
M  V30 4 N -1.200 0.000 0.0 0
M  V30 5 H -1.800 0.800 0.0 0
M  V30 6 H -1.800 -0.800 0.0 0
M  V30 END ATOM
M  V30 BEGIN BOND
M  V30 1 1 1 2
M  V30 2 1 2 3
M  V30 3 1 1 4
M  V30 4 1 4 5
M  V30 5 1 4 6
M  V30 END BOND
M  V30 BEGIN SGROUP
M  V30 1 SUP 1 ATOMS=(1 1 2 3 4 5 6) XBONDS=(1 1) LABEL=MCM88 CLASS=LINKER
M  V30 END SGROUP
M  V30 END CTAB
M  V30 END TEMPLATE
M  END
>  <aliasHELM>
MCM88

>  <idtAliases>
base=mcm88Base,ep5=mcm88Ep5,ep3=mcm88Ep3,i=mcm88I

>  <modificationType>
Micromolecule

$$$$
`, { format: 'sdf' })
```
3. Observe that both XUB and MCM88 appear in the Library.

**Actual behavior**
Monomers with disallowed modificationType (e.g. Unknown base, Micromolecule) are successfully added and displayed in the Library.
No validation or restriction is triggered on the API or UI level.

**Expected behavior**
The following modification types must not be accepted nor shown in the Library:
Ambiguous mixed peptide
Unknown peptide
Ambiguous alternative sugar
Ambiguous mixed sugar
Unknown sugar
Ambiguous mixed DNA base
Ambiguous mixed RNA base
Ambiguous mixed base
Unknown base
Ambiguous alternative phosphate
Ambiguous mixed phosphate
Unknown phosphate
Unknown unsplit nucleotide
Ambiguous alternative CHEM
Ambiguous mixed CHEM
Unknown CHEM
Unknown monomer
Molecule
Micromolecule

**Screenshots**

Image
Image

**Desktop:**
 - OS: Windows 11
 - Browser Chrome
 Version 140.0.7339.128 (Official Build) (64-bit)

**Ketcher version** 
[Version 3.9.0-rc.1]
**Indigo version** 
[Version 1.37.0-rc.1]

This bug was found while testing: https://github.com/epam/ketcher/issues/7674

### PR DESCRIPTION
**Pull request by mariam-khutuashvili**
## How the feature works? / How did you fix the issue?
(Screenshots, videos, or GIFs, if applicable)
 Description

  Monomers imported via updateMonomersLibrary / replaceMonomersLibrary whose modification type is an unknown, ambiguous, or molecule marker (Unknown base, Micromolecule, Ambiguous mixed base, … — the 19 values listed in the issue) are now rejected: they are skipped on load, logged via KetcherLogger, and not added to or shown in the library. Valid monomers in the same import still load (consistent with epam/Indigo#3755).

  Changes

  - ketcher-core/src/utilities/monomers.ts: added DISALLOWED_MONOMER_MODIFICATION_TYPES (the 19 values, as const), DISALLOWED_MODIFICATION_TYPE_ERROR_MESSAGE, and getDisallowedModificationTypes() (returns the offending types).
  - ketcher-core/src/application/editor/Editor.ts: added a check at the start of the updateMonomersLibrary per-monomer loop — if a monomer's props.modificationTypes contains a disallowed value, it is skipped and logged. This gates both updateMonomersLibrary and replaceMonomersLibrary.
  - Tests: getDisallowedModificationTypes unit-tested against all 19 values plus mixed/empty inputs; Editor integration tests for reject (Unknown base, Micromolecule), skip-disallowed-but-load-valid, and accept-allowed-type.

  Testing

  - npm run test --workspace=ketcher-core (prettier, eslint, types, circular-deps, jest — 393 tests incl. the new ones): green.
  - npm run test:types (all workspaces): green.
  - Indigo round-trip verified directly (indigo-ketcher v1.44 in Node): an SDF <modificationTypes> field carrying a disallowed value converts to KET props.modificationTypes, which the validation then rejects — confirming the full SDF → conversion → validation chain.

  Notes for reviewers / QA

  - The validation targets the canonical props.modificationTypes field (the value Indigo produces from a monomer's modification-type data field), so it applies regardless of import path.
  - Current Indigo (v1.44) reads/writes this as the plural data field <modificationTypes>. The issue's example SDF predates the current Indigo: it uses the singular <modificationType> field (which current Indigo ignores) and that exact SDF also no longer parses on the current Indigo. To reproduce on the current build, use an SDF with the plural <modificationTypes> field (or the KET format) carrying a disallowed value — not the issue's literal example.

## Check list
- [ ] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [ ] PR is linked with the issue
- [ ] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [ ] reviewers are notified about the pull request

### CODE REVIEW

File: packages/ketcher-core/src/utilities/monomers.ts [hunk 1]
```diff
@@ -22,6 +22,52 @@ export const MONOMER_GROUP_TEMPLATE_NAME_MAX_LENGTH = 200;
...
+ */
+export function getDisallowedModificationTypes(
+  modificationTypes?: string[],
+): string[] {
+  return (
+    modificationTypes?.filter((type) =>
+      disallowedModificationTypesSet.has(type),
+    ) ?? []
+  );
+}
```
[the remark this answers is not in the document]
Comment (mariam-khutuashvili) [reply packages/ketcher-core/src/utilities/monomers.ts:77]: Added an Array.isArray guard to getDisallowedModificationTypes so a malformed, non-array modificationTypes (the value comes from parsed, untrusted library JSON) returns an empty array instead of throwing. As you noted, a bare string like "modificationTypes": "Unknown base" would otherwise throw a TypeError inside the per-monomer forEach in updateMonomersLibrary and abort the whole chunk, silently dropping the valid monomers after it. Also added a unit test covering a bare string and null, and documented the rationale in the function's JSDoc.

File: packages/ketcher-core/__tests__/application/editor/Editor.test.ts [hunk 2]
```diff
@@ -635,6 +635,169 @@ describe('CoreEditor', () => {
...
+      };
+
+      const initialLibrarySize = editor.monomersLibrary.length;
+      editor.updateMonomersLibrary(JSON.stringify(monomerWithDisallowedType));
+
+      expect(errorSpy).toHaveBeenCalledWith(
+        expect.stringContaining(
+          'Monomers with an unknown, ambiguous, or molecule modification type cannot be added to the library.',
+        ),
+      );
```
[the remark this answers is not in the document]
Comment (mariam-khutuashvili) [reply packages/ketcher-core/__tests__/application/editor/Editor.test.ts:1022]: (a) Added a parsed-JSON assertion to the two reject cases (Unknown base + Micromolecule): I capture editor.monomersLibraryParsedJson?.root.templates.length before the call and assert it's unchanged after — matching the pattern the neighboring monomer-group-template tests already use. So a regression that pushes a rejected template's $ref into the parsed-JSON side-table now fails.

(b) One correction: there's no standalone replaceMonomersLibrary method. The replace entry point is the second argument of initializeMonomersLibraryFromKetcher(update, replace), which clears the library and then delegates to updateMonomersLibrary (where the validation lives). I added a test through that real path — stubbing ketcher.ensureMonomersLibraryDataInKetFormat as a passthrough — asserting a disallowed monomer is rejected via the replace path too, which locks in that contract and exercises initializeMonomersLibraryFromKetcher directly.

**Review by svvald — APPROVED** [review]
